ouroboros-consensus-test/bench-mempool/Main.hs

```diff
@@ -18,37 +33,95 @@ import           Bench.Consensus.MempoolWithMockedLedgerItf
+                --
+                -- error reported here https://hackage.haskell.org/package/tasty-bench-0.3.3/docs/Test-Tasty-Bench.html#v:env
+                (\ ~(mempool, txs) -> bgroup (show n <> " transactions") [
+                    bench    "benchmark"     $ nfIO $ run        mempool txs
+                  , testCase "test"          $        testAddTxs mempool txs
+                  , testCase "txs length"    $ length txs @?= n
+                  ]
+                )
+              where
+                testAddTxs mempool txs = do
+                    run mempool txs
+                    mempoolTxs <- getTxs mempool
+                    mempoolTxs @?= txsAddedInCmds txs
+
+    parseBenchmarkResults csvFilePath = do
+        csvData <- BL.readFile csvFilePath
+        case Csv.decode Csv.HasHeader csvData of
+          Left err   -> die err
+          Right rows -> pure rows
+
+    -- Output the mempool benchmark results as a JSON file, which conforms to
+    -- the input expected by
+    -- https://github.com/benchmark-action/github-action-benchmark
+    convertCsvRowsToJsonObjects rows outFilePath =
+      encodeFile outFilePath $ fmap convertRowToJsonObject rows
       where
-        benchAddNTxs n =
-          env (let txs = mkNTryAddTxs n in fmap (, txs) (openMempoolWithCapacipyFor txs))
-              -- The irrefutable pattern was added to avoid the
-              --
-              -- > Unhandled resource. Probably a bug in the runner you're using.
-              --
-              -- error reported here https://hackage.haskell.org/package/tasty-bench-0.3.3/docs/Test-Tasty-Bench.html#v:env
-              (\ ~(mempool, txs) -> bgroup (show n) [
-                  bench    "benchmark"     $ nfIO $ run        mempool txs
-                , testCase "test"          $        testAddTxs mempool txs
-                , testCase "txs length"    $ length txs @?= n
-                ]
-              )
+        convertRowToJsonObject (name:mean:_) =
+          object [ "name"  .= adjustName name
+                 , "value" .= adjustedMean
+                 , "unit"  .= unit
+                 ]
           where
-            testAddTxs mempool txs = do
-                run mempool txs
-                mempoolTxs <- getTxs mempool
-                mempoolTxs @?= txsAddedInCmds txs
+            adjustName = Text.replace "."          " "
+                       . Text.replace ".benchmark" ""
+
+            adjustedMean :: Integer
+            (adjustedMean, unit) = first round
+                                 $ convertPicosecondsWithUnit
+                                 $ fromInteger
+                                 $ textToInt mean
+              where
+                textToInt = either error fst . Text.Read.decimal
+
+            -- Convert a number of picoseconds to the largest time unit that
+            -- makes the conversion greater or equal than one.
+            convertPicosecondsWithUnit :: Double -> (Double, String)
```

Comment:
Out of curiosity, why is this conversion needed? To scale the benchmarks to the best unit?

Reply:
Right. This is what `Tasty.Benchmark` seems to be doing as well.